File: crates/minos-tui/src/agent_route.rs
use minos_agent_runtime::SessionState;
use minos_domain::AgentName;

/// Host agent profile fields needed for @-routing / mention insert (desktop parity).
#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) struct MentionProfile {
    pub id: String,
    pub name: String,
    pub runtime_agent: AgentName,
    pub updated_at_ms: i64,
}

/// Resolved @-route target.
/// - bare agent: `profile_id` unset — **reuse** open top-level session if any; else start with newest profile
/// - agent#short: continue that session (`profile_id` unset)
/// - profile mention: `profile_id` + agent from profile.runtime_agent (always new session)
#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) struct AgentRouteTarget {
    pub(crate) agent: AgentName,
    pub(crate) session_short_id: Option<String>,
    /// Explicit host agent profile id when routed via @ProfileName or @p/<id>.
    pub(crate) profile_id: Option<String>,
}

pub(crate) fn parse_agent_routing(
    text: &str,
    profiles: &[MentionProfile],
) -> Option<(AgentRouteTarget, String)> {
    let trimmed = text.trim_start();
    let rest = trimmed.strip_prefix('@')?;
    let split_at = rest
        .char_indices()
        .find(|(_, ch)| ch.is_whitespace())
        .map(|(index, _)| index)
        .unwrap_or(rest.len());
    let target = parse_agent_route_target(&rest[..split_at], profiles)?;
    let body = rest[split_at..].trim_start().to_owned();
    Some((target, body))
}

/// Parse a single route token (no leading `@`).
///
/// Resolution order (desktop parity):
/// 1. `p/<id>` or `p:<id>` → profile by id
/// 2. `agent#short` → continue session
/// 3. known runtime agent name → bare agent
/// 4. unique profile name (case-insensitive) → profile
pub(crate) fn parse_agent_route_target(
    value: &str,
    profiles: &[MentionProfile],
) -> Option<AgentRouteTarget> {
    let raw = value.trim();
    if raw.is_empty() {
        return None;
    }

    // Explicit profile id form: p/<id> or p:<id>
    if let Some(rest) = raw
        .strip_prefix("p/")
        .or_else(|| raw.strip_prefix("P/"))
        .or_else(|| raw.strip_prefix("p:"))
        .or_else(|| raw.strip_prefix("P:"))
    {
        let id = rest.trim();
        if id.is_empty() {
            return None;
        }
        let profile = profiles.iter().find(|p| p.id == id)?;
        return Some(AgentRouteTarget {
            agent: profile.runtime_agent,
            session_short_id: None,
            profile_id: Some(profile.id.clone()),
        });
    }

    let (agent_part, session_short_id) = match raw.split_once('#') {
        Some((agent_name, session_short_id)) if !session_short_id.is_empty() => {
            (agent_name, Some(session_short_id.to_owned()))
        }
        Some(_) => return None,
        None => (raw, None),
    };

    // Runtime agents win over same-named profiles (colliding profiles use @p/id).
    if let Some(agent) = parse_agent_name(agent_part) {
        return Some(AgentRouteTarget {
            agent,
            session_short_id,
            profile_id: None,
        });
    }

    // Profile by unique name (case-insensitive). Ignore when #session form.
    if session_short_id.is_some() {
        return None;
    }
    let key = normalize_profile_name(agent_part);
    if key.is_empty() {
        return None;
    }
    let name_matches: Vec<&MentionProfile> = profiles
        .iter()
        .filter(|p| normalize_profile_name(&p.name) == key)
        .collect();
    if name_matches.len() != 1 {
        return None;
    }
    let profile = name_matches[0];
    Some(AgentRouteTarget {
        agent: profile.runtime_agent,
        session_short_id: None,
        profile_id: Some(profile.id.clone()),
    })
}

pub(crate) fn parse_agent_name(value: &str) -> Option<AgentName> {
    let normalized = value.to_ascii_lowercase();
    AgentName::all()
        .iter()
        .copied()
        .find(|agent| agent.bin_name() == normalized.as_str())
}

pub(crate) fn normalize_profile_name(name: &str) -> String {
    name.trim().to_ascii_lowercase()
}

/// Whether a profile display name can be a single `@Name` route token.
/// Whitespace breaks token split; `#` is agent#session form; `@` nests mentions.
pub(crate) fn is_profile_name_clean_token(name: &str) -> bool {
    let n = name.trim();
    if n.is_empty() {
        return false;
    }
    !n.chars().any(|ch| ch.is_whitespace() || ch == '#' || ch == '@')
}

/// Unique among profiles + runtime agent names (case-insensitive).
pub(crate) fn is_profile_name_unique(name: &str, profiles: &[MentionProfile]) -> bool {
    let key = normalize_profile_name(name);
    if key.is_empty() {
        return false;
    }
    if AgentName::all()
        .iter()
        .any(|a| a.bin_name().eq_ignore_ascii_case(&key))
    {
        return false;
    }
    let hits = profiles
        .iter()
        .filter(|p| normalize_profile_name(&p.name) == key)
        .count();
    hits == 1
}

/// Insert token for a profile: `@Name ` when unique **and** a clean token, else `@p/<id> `.
pub(crate) fn profile_mention_insert(profile: &MentionProfile, profiles: &[MentionProfile]) -> String {
    if is_profile_name_clean_token(&profile.name) && is_profile_name_unique(&profile.name, profiles)
    {
        format!("@{} ", profile.name.trim())
    } else {
        format!("@p/{} ", profile.id)
    }
}

/// Newest host profile id for a runtime (by `updated_at_ms`), if any.
pub(crate) fn newest_profile_id_for_agent(
    profiles: &[MentionProfile],
    agent: AgentName,
) -> Option<String> {
    profiles
        .iter()
        .filter(|p| p.runtime_agent == agent)
        .max_by_key(|p| p.updated_at_ms)
        .map(|p| p.id.clone())
}

/// First up-to-8 bytes of a session id for display / mention tokens.
///
/// Session ids are ASCII hex (or similar); this returns a borrowed slice so
/// render paths never allocate.
pub(crate) fn short_session_id(session_id: &str) -> &str {
    let mut end = session_id.len().min(8);
    while end > 0 && !session_id.is_char_boundary(end) {
        end -= 1;
    }
    &session_id[..end]
}

pub(crate) fn thread_can_receive_message(state: &SessionState) -> bool {
    !matches!(state, SessionState::Closed { .. })
}

/// Fields needed to pick a reusable top-level session (desktop bare-`@agent` parity).
#[derive(Debug, Clone, Copy)]
pub(crate) struct SessionReuseCandidate<'a> {
    pub session_id: &'a str,
    pub agent: AgentName,
    pub parent_session_id: Option<&'a str>,
    pub state: &'a SessionState,
    pub last_ts_ms: i64,
}

/// Most recent top-level, non-closed session for `agent`, if any.
///
/// Matches desktop send use-case: bare `@agent` reuses before starting a new run.
/// Subagents (`parent_session_id` set) and closed sessions are excluded.
pub(crate) fn pick_reusable_session_id(
    candidates: &[SessionReuseCandidate<'_>],
    agent: AgentName,
) -> Option<String> {
    candidates
        .iter()
        .filter(|c| c.agent == agent)
        .filter(|c| c.parent_session_id.is_none())
        .filter(|c| thread_can_receive_message(c.state))
        .max_by_key(|c| c.last_ts_ms)
        .map(|c| c.session_id.to_owned())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn profile(id: &str, name: &str, agent: AgentName, updated: i64) -> MentionProfile {
        MentionProfile {
            id: id.into(),
            name: name.into(),
            runtime_agent: agent,
            updated_at_ms: updated,
        }
    }

    #[test]
    fn short_session_id_returns_borrowed_prefix() {
        let id = "abcdef0123456789";
        assert_eq!(short_session_id(id), "abcdef01");
        assert_eq!(short_session_id("abc"), "abc");
        assert_eq!(short_session_id(""), "");
    }

    #[test]
    fn short_session_id_respects_utf8_char_boundary() {
        // "abcdef" (6) + 你 (3 bytes at 6..=8) → raw cut at 8 is mid-char.
        let id = "abcdef你g";
        let short = short_session_id(id);
        assert_eq!(short, "abcdef");
        assert!(id.is_char_boundary(short.len()));
    }

    #[test]
    fn parses_bare_agent_and_session_short() {
        let (target, body) = parse_agent_routing("@codex hello", &[]).unwrap();
        assert_eq!(target.agent, AgentName::Codex);
        assert!(target.session_short_id.is_none());
        assert!(target.profile_id.is_none());
        assert_eq!(body, "hello");

        let (target, body) = parse_agent_routing("@grok#abcd1234 continue", &[]).unwrap();
        assert_eq!(target.agent, AgentName::Grok);
        assert_eq!(target.session_short_id.as_deref(), Some("abcd1234"));
        assert!(target.profile_id.is_none());
        assert_eq!(body, "continue");
    }

    #[test]
    fn parses_unique_profile_name() {
        let profiles = vec![profile(
            "profile-research",
            "ResearchGrok",
            AgentName::Grok,
            10,
        )];
        let (target, body) = parse_agent_routing("@ResearchGrok dig in", &profiles).unwrap();
        assert_eq!(target.agent, AgentName::Grok);
        assert_eq!(target.profile_id.as_deref(), Some("profile-research"));
        assert_eq!(body, "dig in");
    }

    #[test]
    fn parses_profile_by_id_token() {
        let profiles = vec![profile("profile-research", "ResearchGrok", AgentName::Grok, 1)];
        let (target, _) = parse_agent_routing("@p/profile-research go", &profiles).unwrap();
        assert_eq!(target.profile_id.as_deref(), Some("profile-research"));
        assert_eq!(target.agent, AgentName::Grok);
    }

    #[test]
    fn runtime_name_wins_over_same_named_profile() {
        let profiles = vec![profile("profile-codex", "codex", AgentName::Grok, 1)];
        let (target, _) = parse_agent_routing("@codex hi", &profiles).unwrap();
        assert_eq!(target.agent, AgentName::Codex);
        assert!(target.profile_id.is_none());
    }

    #[test]
    fn ambiguous_profile_name_does_not_parse() {
        let profiles = vec![
            profile("p1", "Helper", AgentName::Grok, 1),
            profile("p2", "Helper", AgentName::Codex, 2),
        ];
        assert!(parse_agent_routing("@Helper hi", &profiles).is_none());
    }

    #[test]
    fn profile_mention_insert_uses_p_id_when_name_collides_with_runtime() {
        let profiles = vec![profile("p1", "codex", AgentName::Grok, 1)];
        let insert = profile_mention_insert(&profiles[0], &profiles);
        assert_eq!(insert, "@p/p1 ");
    }

    #[test]
    fn newest_profile_id_picks_latest_updated() {
        let profiles = vec![
            profile("old", "A", AgentName::Grok, 1),
            profile("new", "B", AgentName::Grok, 99),
            profile("other", "C", AgentName::Codex, 1000),
        ];
        assert_eq!(
            newest_profile_id_for_agent(&profiles, AgentName::Grok).as_deref(),
            Some("new")
        );
        assert_eq!(
            newest_profile_id_for_agent(&profiles, AgentName::Claude),
            None
        );
    }

    #[test]
    fn rejects_unclean_profile_token_chars_for_name_uniqueness_insert() {
        assert!(!is_profile_name_clean_token("has space"));
        assert!(!is_profile_name_clean_token("has#hash"));
        assert!(!is_profile_name_clean_token("has@at"));
        assert!(is_profile_name_clean_token("ResearchGrok"));
    }

    #[test]
    fn pick_reusable_session_prefers_latest_open_top_level() {
        let idle = SessionState::Idle;
        let closed = SessionState::Closed {
            reason: minos_agent_runtime::CloseReason::UserClose,
        };
        let candidates = [
            SessionReuseCandidate {
                session_id: "old-codex",
                agent: AgentName::Codex,
                parent_session_id: None,
                state: &idle,
                last_ts_ms: 10,
            },
            SessionReuseCandidate {
                session_id: "new-codex",
                agent: AgentName::Codex,
                parent_session_id: None,
                state: &idle,
                last_ts_ms: 99,
            },
            SessionReuseCandidate {
                session_id: "child",
                agent: AgentName::Codex,
                parent_session_id: Some("new-codex"),
                state: &idle,
                last_ts_ms: 200,
            },
            SessionReuseCandidate {
                session_id: "closed",
                agent: AgentName::Codex,
                parent_session_id: None,
                state: &closed,
                last_ts_ms: 300,
            },
            SessionReuseCandidate {
                session_id: "grok",
                agent: AgentName::Grok,
                parent_session_id: None,
                state: &idle,
                last_ts_ms: 500,
            },
        ];
        assert_eq!(
            pick_reusable_session_id(&candidates, AgentName::Codex).as_deref(),
            Some("new-codex"),
        );
        assert_eq!(
            pick_reusable_session_id(&candidates, AgentName::Claude),
            None,
        );
    }
}

File: apps/desktop/src/shared/lib/agent-route.ts
/** Parse `@agent` / `@agent#short` / `@profile` routing (aligned with minos-tui agent_route.rs + desktop profiles). */

/**
 * Offline parse fallback for `@agent` tokens when CLI inventory is empty.
 * Production mention rows come from `buildAgentMentionOptions(clis, …)` using
 * daemon `list_clis`. Do **not** treat this as the runtime capability catalog —
 * see `features/agents/AGENTS.md` and domain `AgentName`.
 */
export const KNOWN_AGENTS = [
  "codex",
  "claude",
  "gemini",
  "opencode",
  "grok",
] as const;

export type KnownAgent = (typeof KNOWN_AGENTS)[number];

/**
 * Resolved @-route target.
 * - bare agent / agent#short: profileId unset (start uses newest profile default)
 * - profile mention: profileId + agent from profile.runtime_agent
 */
export type AgentRouteTarget = {
  agent: KnownAgent;
  sessionShortId?: string;
  /** Explicit host agent profile id when routed via @ProfileName or @p/<id>. */
  profileId?: string;
};

/** First up-to-8 chars of a session id (TUI `short_session_id` parity). */
export function shortSessionId(sessionId: string): string {
  let end = Math.min(8, sessionId.length);
  // Avoid splitting multi-byte code units at the cut (rare for hex ids).
  while (end > 0 && (sessionId.charCodeAt(end - 1) & 0xfc00) === 0xdc00) {
    end -= 1;
  }
  return sessionId.slice(0, end);
}

export function parseAgentName(value: string): KnownAgent | null {
  const normalized = value.toLowerCase();
  return (KNOWN_AGENTS as readonly string[]).includes(normalized)
    ? (normalized as KnownAgent)
    : null;
}

export type MentionProfile = {
  id: string;
  name: string;
  runtimeAgent: string;
};

/** Normalize a profile name for uniqueness / match (case-insensitive, trimmed). */
export function normalizeProfileName(name: string): string {
  return name.trim().toLowerCase();
}

/**
 * Whether a profile display name can be a single `@Name` route token.
 * Whitespace breaks token split; `#` is agent#session form; `@` nests mentions.
 * Daemon create/update rejects these; this is defense-in-depth for insert/parse.
 */
export function isProfileNameCleanToken(name: string): boolean {
  const n = name.trim();
  if (!n) return false;
  return !/[\s#@]/.test(n);
}

/**
 * Validate profile display name for create/update forms (daemon-aligned).
 * Returns an error message, or null when valid.
 */
export function validateProfileName(name: string): string | null {
  const n = name.trim();
  if (!n) return "Name is required";
  if (/[\s#@]/.test(n)) {
    return "Name cannot contain spaces, #, or @";
  }
  return null;
}

/**
 * Whether a profile display name is unique among profiles + runtime agent names.
 * Collisions force `@p/<id>` insert tokens so parse is unambiguous.
 */
export function isProfileNameUnique(
  name: string,
  profiles: readonly MentionProfile[],
  runtimeAgents: readonly string[] = KNOWN_AGENTS,
): boolean {
  const key = normalizeProfileName(name);
  if (!key) return false;
  if (runtimeAgents.some((a) => a.toLowerCase() === key)) return false;
  let hits = 0;
  for (const p of profiles) {
    if (normalizeProfileName(p.name) === key) hits += 1;
    if (hits > 1) return false;
  }
  return hits === 1;
}

/**
 * Insert token for a profile: `@Name ` when unique **and** a clean token,
 * else `@p/<id> ` (defense in depth for legacy / unsafe names).
 */
export function profileMentionInsert(
  profile: MentionProfile,
  profiles: readonly MentionProfile[],
  runtimeAgents: readonly string[] = KNOWN_AGENTS,
): string {
  if (
    isProfileNameCleanToken(profile.name) &&
    isProfileNameUnique(profile.name, profiles, runtimeAgents)
  ) {
    // Preserve user casing from profile.name for readability.
    return `@${profile.name.trim()} `;
  }
  return `@p/${profile.id} `;
}

/**
 * Parse a single route token (no leading `@`).
 *
 * Resolution order:
 * 1. `p/<id>` or `p:<id>` → profile by id (requires profiles list)
 * 2. `agent#short` → continue session
 * 3. known runtime agent name → bare agent
 * 4. unique profile name (case-insensitive) → profile
 */
export function parseAgentRouteTarget(
  value: string,
  profiles: readonly MentionProfile[] = [],
): AgentRouteTarget | null {
  const raw = value.trim();
  if (!raw) return null;

  // Explicit profile id form: p/<id> or p:<id>
  const pidMatch = /^p[/:](.+)$/i.exec(raw);
  if (pidMatch) {
    const id = pidMatch[1]?.trim();
    if (!id) return null;
    const profile = profiles.find((p) => p.id === id);
    if (!profile) {
      // Still accept token shape; agent unknown until profiles load — fail closed.
      return null;
    }
    const agent = parseAgentName(profile.runtimeAgent);
    if (!agent) return null;
    return { agent, profileId: profile.id };
  }

  const [agentPart, shortPart] = raw.split("#");
  if (shortPart !== undefined && shortPart.length === 0) return null;

  // Runtime agents win over same-named profiles (colliding profiles use @p/id).
  const agent = parseAgentName(agentPart ?? "");
  if (agent) {
    return {
      agent,
      sessionShortId: shortPart || undefined,
    };
  }

  // Profile by unique name (case-insensitive). Ignore when #session form.
  if (shortPart !== undefined) return null;
  const key = normalizeProfileName(agentPart ?? "");
  if (!key) return null;
  const nameMatches = profiles.filter(
    (p) => normalizeProfileName(p.name) === key,
  );
  if (nameMatches.length !== 1) return null;
  const profile = nameMatches[0]!;
  const profileAgent = parseAgentName(profile.runtimeAgent);
  if (!profileAgent) return null;
  return { agent: profileAgent, profileId: profile.id };
}

/**
 * `@codex hello` → { target: { agent: "codex" }, prompt: "hello", messageBody: "@codex hello" }
 * `@ResearchGrok hello` → profile target when name unique
 * `@p/profile-uuid hello` → profile by id
 * plain text → null (caller may fall back to default agent)
 *
 * Routing semantics:
 * - `@agent prompt` → start/reuse session for that agent (newest profile default on create)
 * - `@agent#short prompt` → continue an existing session
 * - `@ProfileName` / `@p/<id>` → **new** session with explicit profile_id
 */
export function parseAgentRouting(
  text: string,
  profiles: readonly MentionProfile[] = [],
): { target: AgentRouteTarget; prompt: string; messageBody: string } | null {
  return parseAllAgentRoutings(text, profiles)[0] ?? null;
}

export type AgentRouting = {
  target: AgentRouteTarget;
  prompt: string;
  messageBody: string;
};

/**
 * Multi-@ fan-out: every unique agent mentioned in the body (appearance order).
 * Shared prompt is the body with leading agent @tokens stripped when the
 * message starts with @; otherwise the full text is the prompt for each agent.
 */
export function parseAllAgentRoutings(
  text: string,
  profiles: readonly MentionProfile[] = [],
): AgentRouting[] {
  const messageBody = text.trimEnd();
  const trimmed = text.trimStart();
  if (!trimmed) return [];

  const routes: AgentRouting[] = [];
  const seen = new Set<string>();

  // Walk @tokens anywhere in the body (not only leading).
  const re = /(^|[\s([{"'`])@([^\s@]+)/g;
  let match: RegExpExecArray | null;
  while ((match = re.exec(trimmed)) !== null) {
    const token = match[2] ?? "";
    if (!token) continue;
    const target = parseAgentRouteTarget(token, profiles);
    if (!target) continue;
    const key = target.profileId
      ? `p:${target.profileId}`
      : target.sessionShortId
        ? `${target.agent}#${target.sessionShortId}`
        : target.agent;
    if (seen.has(key)) continue;
    // Dedup by agent runtime for bare @agent (one session per agent per turn).
    const agentKey = target.agent;
    if (!target.sessionShortId && !target.profileId && seen.has(`a:${agentKey}`)) {
      continue;
    }
    seen.add(key);
    if (!target.sessionShortId && !target.profileId) {
      seen.add(`a:${agentKey}`);
    }
    routes.push({ target, prompt: "", messageBody });
  }

  if (routes.length === 0) return [];

  // Shared prompt: strip leading consecutive agent @tokens.
  let prompt = trimmed;
  while (prompt.startsWith("@")) {
    const rest = prompt.slice(1);
    const splitAt = [...rest].findIndex((ch) => /\s/.test(ch));
    const token = splitAt === -1 ? rest : rest.slice(0, splitAt);
    if (!parseAgentRouteTarget(token, profiles)) break;
    prompt = splitAt === -1 ? "" : rest.slice(splitAt).trimStart();
  }
  // Also strip mid-body agent mentions when message is pure multi-@ + instruction
  // is already handled by leading strip; keep full body if no leading @.
  const sharedPrompt = prompt.trim() || messageBody;

  return routes.map((r) => ({ ...r, prompt: sharedPrompt }));
}

/** Active @-token at cursor for autocomplete (TUI-style). */
export function mentionQueryAtCursor(
  text: string,
  cursor: number,
): { start: number; query: string } | null {
  const before = text.slice(0, cursor);
  const at = before.lastIndexOf("@");
  if (at < 0) return null;
  if (at > 0 && !/\s/.test(before[at - 1] ?? " ")) return null;
  const token = before.slice(at + 1);
  if (/\s/.test(token)) return null;
  return { start: at, query: token };
}

export type MentionOption = {
  id: string;
  label: string;
  hint: string;
  insert: string;
  disabled: boolean;
};

export type MentionCli = {
  agent: string;
  installed: boolean;
  status: string;
};

export type MentionSession = {
  id: string;
  agent: string;
  shortId: string;
  status: string;
  parentId?: string | null;
};

export type BuildAgentMentionOptionsArgs = {
  query: string;
  clis: readonly MentionCli[];
  sessions: readonly MentionSession[];
  profiles?: readonly MentionProfile[];
  /**
   * Conversation roster (membership SSOT). When provided, only these runtime
   * agents (and profiles/sessions for them) appear. Empty array ⇒ no options.
   * Omit only for offline/unit contexts that intentionally list all CLIs.
   */
  memberAgents?: readonly string[];
  limit?: number;
};

/**
 * TUI-parity @-picker rows (membership-gated when `memberAgents` is set):
 * 1. Member runtimes as bare `@agent` → start (newest profile default)
 * 2. Host profiles for member runtimes as `@ProfileName` / `@p/id`
 * 3. Existing open member sessions as `@agent#short` → continue
 */
export function buildAgentMentionOptions(
  queryOrArgs: string | BuildAgentMentionOptionsArgs,
  clisArg?: readonly MentionCli[],
  sessionsArg?: readonly MentionSession[],
  profilesArg: readonly MentionProfile[] = [],
  limitArg = 16,
): MentionOption[] {
  // Support both positional (legacy tests) and object form.
  const args: BuildAgentMentionOptionsArgs =
    typeof queryOrArgs === "string"
      ? {
          query: queryOrArgs,
          clis: clisArg ?? [],
          sessions: sessionsArg ?? [],
          profiles: profilesArg,
          limit: limitArg,
        }
      : queryOrArgs;

  const query = args.query;
  const clis = args.clis;
  const sessions = args.sessions;
  const profiles = args.profiles ?? [];
  const limit = args.limit ?? 16;
  const memberFilter =
    args.memberAgents === undefined
      ? null
      : new Set(
          args.memberAgents
            .map((a) => a.trim().toLowerCase())
            .filter((a) => a.length > 0),
        );

  const isMember = (agent: string) =>
    memberFilter === null || memberFilter.has(agent.toLowerCase());

  // Explicit empty roster: nothing is @-able.
  if (memberFilter !== null && memberFilter.size === 0) {
    return [];
  }

  const q = query.toLowerCase();
  const matches = (s: string) => !q || s.toLowerCase().includes(q);

  const memberClis = clis.filter((c) => isMember(c.agent));
  const fromCli = memberClis
    .filter((c) => matches(c.agent) || matches(`@${c.agent}`))
    .map((c) => ({
      id: `new:${c.agent}`,
      label: `@${c.agent}`,
      hint: c.installed ? "new session" : "not installed",
      insert: `@${c.agent} `,
      disabled: !c.installed,
    }));

  const fromKnown =
    fromCli.length > 0
      ? fromCli
      : memberFilter === null
        ? KNOWN_AGENTS.filter((a) => matches(a) || matches(`@${a}`)).map(
            (a) => ({
              id: `new:${a}`,
              label: `@${a}`,
              hint: "new session",
              insert: `@${a} `,
              disabled: false,
            }),
          )
        : // Roster set but CLI inventory empty: still offer bare member tokens.
          [...memberFilter]
            .filter((a) => matches(a) || matches(`@${a}`))
            .map((a) => ({
              id: `new:${a}`,
              label: `@${a}`,
              hint: "new session",
              insert: `@${a} `,
              disabled: false,
            }));

  const runtimeNames = [
    ...new Set([
      ...KNOWN_AGENTS,
      ...clis.map((c) => c.agent),
      ...(memberFilter ? [...memberFilter] : []),
    ]),
  ];

  const fromProfiles = profiles
    .filter((p) => isMember(p.runtimeAgent))
    .filter(
      (p) =>
        matches(p.name) ||
        matches(p.runtimeAgent) ||
        matches(`@${p.name}`) ||
        matches(`p/${p.id}`),
    )
    .map((p) => {
      const useName =
        isProfileNameCleanToken(p.name) &&
        isProfileNameUnique(p.name, profiles, runtimeNames);
      const label = useName
        ? `@${p.name.trim()}`
        : `@p/${p.id.slice(0, 12)}…`;
      return {
        id: `profile:${p.id}`,
        label,
        hint: `profile · ${p.runtimeAgent}`,
        insert: profileMentionInsert(p, profiles, runtimeNames),
        disabled: false,
      };
    });

  const fromSessions = sessions
    .filter((s) => isMember(s.agent))
    .filter((s) => !s.parentId)
    .filter((s) => s.status !== "done" && s.status !== "failed")
    .filter(
      (s) =>
        matches(s.agent) ||
        matches(s.shortId) ||
        matches(`@${s.agent}#${s.shortId}`),
    )
    .map((s) => ({
      id: `sess:${s.id}`,
      label: `@${s.agent}#${s.shortId}`,
      hint: `continue · ${s.status}`,
      insert: `@${s.agent}#${s.shortId} `,
      disabled: false,
    }));

  return [...fromKnown, ...fromProfiles, ...fromSessions].slice(0, limit);
}
